codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,297 @@
+                // be allocated bytes when the parent stream can't utilize them. This may be as a result of the
+                // stream's flow control window being 0.
+                pState.updateStreamableBytes(pState.streamableBytes, false);
+                return 0;
+            }
+            connectionState.pseudoTime += nsent;
+            return nsent;
+        }
+
+        return distributeToChildren(maxBytes, writer, pState);
+    }
+
+    private int distributeToChildren(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        long oldTotalQueuedWeights = pState.totalQueuedWeights;
+        State iState = pState.poll();
+        State iNextState = pState.peek();
+        try {
+            assert iNextState == null || iNextState.pseudoTime >= iState.pseudoTime : "iNextState.pseudoTime(" +
+                    iNextState.pseudoTime + ") < " + " iState.pseudoTime(" + iState.pseudoTime + ")";
+            int nsent = distribute(iNextState == null ? maxBytes :
+                                    min(maxBytes, (int) ((iNextState.pseudoTime - iState.pseudoTime) *
+                                             ((double) iNextState.stream.weight() / oldTotalQueuedWeights)
+                                             + allocationQuantum)),
+                               writer,
+                               iState);
+            iState.updatePseudoTime(nsent, oldTotalQueuedWeights);
+            return nsent;
+        } finally { // Do in finally to ensure the internal state is not corrupted if an exception is thrown.
+            // The offer operation is delayed until we unroll up the recursive stack, so we don't have to remove from
+            // the priority queue due to a write operation.
+            if (iState.isActiveCountForTree != 0) {
+                pState.offer(iState);
+            }
+        }
+    }
+
+    private State state(Http2Stream stream) {
+        return stream.getProperty(stateKey);
+    }
+
+    /**
+     * For testing only!
+     */
+    int streamableBytes(Http2Stream stream) {
+        return state(stream).streamableBytes;
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State implements PriorityQueueNode<State> {
+        final Http2Stream stream;
+        private final Queue<State> queue;
+        int streamableBytes;
+        int isActiveCountForTree;
+        private int priorityQueueIndex = INDEX_NOT_IN_QUEUE;
+        long pseudoTime;
+        long totalQueuedWeights;
+        boolean isActive;
+
+        State(Http2Stream stream) {
+            this(stream, 0, connectionState.pseudoTime);
+        }
+
+        State(Http2Stream stream, int initialSize, long pseudoTime) {
+            this.stream = stream;
+            queue = new PriorityQueue<State>(initialSize);
+            this.pseudoTime = pseudoTime;
```

Comment:
All the `pseudoTime`s can be initialized to `0`. `pseudoTime` should be "initialized" just initially adding to a `queue`, because the `pseudoTime` is supposed to represent when the ideal NIC would have sent data for the stream, but even the ideal NIC can't sent data before the stream has actually provided it. Put another way, streams aren't substantially boosted for the time they _didn't_ send; if a stream was idle for an hour, it doesn't get credit to use that bandwidth later, as that would starve other streams.